src/relax/transform/to_mixed_precision.cc

```diff
@@ -0,0 +1,297 @@
+ * under the License.
+ */
+/*!
+ * \file src/relax/transform/to_mixed_precision.cc
+ * \brief Automatic mixed precision pass.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include "../op/make_op.h"
+
+namespace tvm {
+namespace relax {
+
+enum MixedTypePolicy : int {
+  MIXED_PRECISION_ALWAYS = 0,
+  MIXED_PRECISION_FOLLOW = 1,
+  MIXED_PRECISION_NEVER = 2
+};
+
+// Return array is of type : [MixedTypePolicy (int), bool, Call]
+// Call is a call node, out_dtype_str is the expected output_dtype string
+using FMixedPrecision = runtime::TypedPackedFunc<Array<ObjectRef>(
+    const Call& call_node, const std::string& out_dtype_str)>;
+
+class ToMixedPrecisionMutator : public ExprMutator {
+ public:
+  explicit ToMixedPrecisionMutator(DLDataType output_dtype) : output_dtype_(output_dtype) {}
+
+  void InitVarMap(const Function& func) {
+    for (const auto& param : func->params) {
+      if (const auto* type = param->checked_type_.as<DynTensorTypeNode>()) {
+        UpdateVarMap(param, type->dtype, param);
+      }
+    }
+  }
+
+  // Visit the use-site of a defined Var
+  Expr VisitExpr_(const VarNode* op) {
+    if (const auto* type = op->checked_type_.as<DynTensorTypeNode>()) {
+      auto it = var_map_.find(GetRef<Var>(op));
+      ICHECK(it != var_map_.end());
+      auto type_it = it->second.find(runtime::DLDataType2String(type->dtype));
+      if (type_it != it->second.end()) {
+        return type_it->second;
+      } else {
+        Var cur_var = it->second.begin()->second;
+        return builder_->Emit(MakeCast(cur_var, type->dtype));
+      }
+    } else {
+      return ExprMutator::VisitExpr_(op);
+    }
+  }
+
+  // Visit the use-site of a defined DataflowVar
+  Expr VisitExpr_(const DataflowVarNode* op) {
+    if (const auto* type = op->checked_type_.as<DynTensorTypeNode>()) {
+      auto it = var_map_.find(GetRef<Var>(op));
+      ICHECK(it != var_map_.end());
+      auto type_it = it->second.find(runtime::DLDataType2String(type->dtype));
+      if (type_it != it->second.end()) {
+        return type_it->second;
+      } else {
+        Var cur_var = it->second.begin()->second;
+        return builder_->Emit(MakeCast(cur_var, type->dtype));
+      }
+    } else {
+      return ExprMutator::VisitExpr_(op);
+    }
+  }
+
+  void VisitBinding_(const VarBindingNode* binding) override {
+    auto emit = [this, binding](const Expr& e, const Var& v) -> Var {
+      Var new_var;
+      if (this->builder_->CurrentBlockIsDataFlow() && !binding->var.as<DataflowVarNode>()) {
+        new_var = this->builder_->EmitOutput(e);
+      } else {
+        new_var = this->builder_->Emit(e);
+      }
+      if (const auto* type = new_var->checked_type_.as<DynTensorTypeNode>()) {
+        this->UpdateVarMap(binding->var, type->dtype, new_var);
+      }
+      return new_var;
+    };
+
+    if (const CallNode* call_node = binding->value.as<CallNode>()) {
```

Comment:
Can we directly do rewrite in VisitExpr(CallNode*?(

Reply:
No.

Reply:
You can return Nested Exprs,

For example, `return Call(Relu, Call(x))`, the caller will normalize it and create new binding. Let me know if that is the main neeed

Reply:
I also need to know which var this call binds to track the type of vars

Reply:
The Var lookup is something that we cannot do indeed, in that case, we can use the second level dispatch VisitBinding_(VarBinding, CallNode*op), which is newly added to the mutator

Reply:
OK I see it. I will refactor it anyway since the current impl doesn't consider the var to be tuple of tensors.